base/services/w32time/w32time.c

```diff
@@ -0,0 +1,418 @@
+    li.QuadPart += (LONGLONG)10000000 * ulTime;
+    ftNew = * (FILETIME *)&li;
+
+    /* Convert back to a system time */
+    if (!FileTimeToSystemTime(&ftNew, &stNew))
+    {
+        return GetLastError();
+    }
+
+    if (!SystemSetTime(&stNew))
+    {
+        return GetLastError();
+    }
+
+    return ERROR_SUCCESS;
+}
+
+
+static DWORD
+GetIntervalSetting(VOID)
+{
+    HKEY hKey;
+    DWORD dwData;
+    DWORD dwSize = sizeof(dwData);
+    LONG lRet;
+
+    dwData = 0;
+    if (RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                      L"SYSTEM\\CurrentControlSet\\Services\\W32Time\\TimeProviders\\NtpClient",
+                      0,
+                      KEY_QUERY_VALUE,
+                      &hKey) == ERROR_SUCCESS)
+    {
+        /* This key holds the update interval in seconds
+         * It is useful for testing to set it to a value of 10 (Decimal)
+         * This will cause the clock to try and update every 10 seconds
+         * So you can change the time and expect it to be set back correctly in 10-20 seconds
+         */
+        lRet = RegQueryValueExW(hKey,
+                                L"SpecialPollInterval",
+                                NULL,
+                                NULL,
+                                (LPBYTE)&dwData,
+                                &dwSize);
+        RegCloseKey(hKey);
+    }
+
+    if (lRet != ERROR_SUCCESS)
+        return 0;
+    else
+        return dwData;
+}
+
+
+int
+SetTime(VOID)
+{
+    ULONG ulTime;
+    LONG lRet;
+    HKEY hKey;
+    DWORD dwIndex = 0;
+    DWORD dwValSize;
+    WCHAR szValName[MAX_VALUE_NAME];
+    WCHAR szData[MAX_VALUE_NAME];
+    DWORD cbName = sizeof(szData);
+    WCHAR szDefault[MAX_VALUE_NAME] = L"";
+    size_t Size;
+
+    DPRINT("Entered SetTime.\n");
+
+    lRet = RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                         L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\DateTime\\Servers",
+                         0,
+                         KEY_QUERY_VALUE,
+                         &hKey);
+    if (lRet != ERROR_SUCCESS)
+    {
+        return 1;
+    }
+
+    while (TRUE)
+    {
+        dwValSize = sizeof(szValName);
+        szValName[0] = UNICODE_NULL;
+        lRet = RegEnumValueW(hKey,
+                             dwIndex,
+                             szValName,
+                             &dwValSize,
+                             NULL,
+                             NULL,
+                             (LPBYTE)szData,
+                             &cbName);
+        if (lRet == ERROR_SUCCESS)
+        {
+            /* Get data from default registry value */
+            if (szValName[0] == UNICODE_NULL) /* if this is the "(Default)" key */
+            {
+                StringCbCopyNW(szDefault, sizeof(szDefault), szData, cbName);
+            }
+            dwIndex++;
+        }
+        else if (lRet != ERROR_MORE_DATA)
+        {
+            break;
+        }
+    }
+
+    dwIndex = 0;
+    while (TRUE)
```

Comment:
I think this similarly can become RegQueryValueEx.

Reply:
Yes, I agree.  See Above comment.